modules/calc/src/main/java/com/opengamma/strata/calc/runner/function/CalculationFunction.java

```diff
@@ -39,4 +61,32 @@
     return Optional.empty();
   }
 
+  /**
+   * Determines the market data requirements that function has to perform its calculations.
+   * <p>
+   * Any market data needed by the {@code calculate} method should be specified.
+   *
+   * @param target  the target of the calculation
+   * @param measures  the set of measures to be calculated
+   * @return the requirements specifying the market data the function needs to perform calculations
+   */
+  public abstract FunctionRequirements requirements(T target, Set<Measure> measures);
+
+  /**
+   * Calculates values of multiple measures for the target using multiple sets of market data.
+   * <p>
+   * The set of measures must only contain measures that the function supports,
+   * as returned by {@link #supportedMeasures()}. The market data must provide at least the
+   * set of data requested by {@link #requirements(CalculationTarget, Set)}.
+   *
+   * @param target  the target of the calculation
+   * @param measures  the set of measures to calculate
+   * @param marketData  the market data to be used in the calculation
+   * @return the read-only map of calculated values, keyed by their measure
+   */
+  public abstract Map<Measure, Result<ScenarioResult<?>>> calculate(
```

Comment:
I don't think the return type is right. At the moment there is no requirement for a function to return a `ScenarioResult`. It's possible to write a function that returns a single result which is an aggregate value for all scenarios.

As a trivial example, it's possible to write a function that returns the maximum PV for a set of scenarios. That would be a single value applying to all scenarios. Constraining the result type to `ScenarioResult` implies the results will always be associated with a particular scenario.

The current return type of `CalculationSingleFunction.execute` is any type, not a subtype of `ScenarioResult`.